src/sidebar/services/tags.js

```diff
@@ -19,17 +19,32 @@ function tags(localStorage) {
   const TAGS_LIST_KEY = 'hypothesis.user.tags.list';
   const TAGS_MAP_KEY = 'hypothesis.user.tags.map';
 
+  const testTags = [];
+  for (let i = 0; i < 1000; i++) {
+    testTags.push(`test-tag-${i}`);
+  }
+
   /**
    * Return a list of tag suggestions matching `query`.
    *
    * @param {string} query
+   * @param {number} limit - Optional limit of the results.
    * @return {Tag[]} List of matching tags
    */
-  function filter(query) {
+  function filter(query, limit = null) {
     const savedTags = localStorage.getObject(TAGS_LIST_KEY) || [];
-
+    //const savedTags = testTags;
+    let resultCount = 0;
     return savedTags.filter(e => {
-      return e.toLowerCase().indexOf(query.toLowerCase()) !== -1;
+      if (e.toLowerCase().indexOf(query.toLowerCase()) !== -1) {
+        if (!limit || resultCount < limit) {
```

Comment:
`!limit` will be true here if it is equal to `0`, which means that `tags.filter(query, 0)` could unexpectedly return a non-empty list.

I wasn't sure if it was safe to assume that `filter` is called with each element in the array once in any particular order, but it turns out that the JS spec does specify the iteration order ([spec link](https://tc39.es/ecma262/#sec-array.prototype.filter)).

Reply:
I think I was just assuming a limit of 0 made no sense so I didn't bother with an extra condition. e.g. a limit of 0 would just allow the entire list, but I can see how that is confusing.